### PR TITLE
feat: implement local puller and pusher

### PR DESCRIPTION
**Pull request by thesayyn**
Take two on https://github.com/google/go-containerregistry/issues/1433

Most of the crane cli surface will be able to work with images on oci-layout.

### CODE REVIEW

File: pkg/v1/local/layer.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package local
```
Comment (imjasonh) [inline pkg/v1/local/layer.go:1]: Is there a reason this shouldn't just go in the layout package? 🤔

File: pkg/.DS_Store
Comment (imjasonh) [inline pkg/.DS_Store:1]: Should remove this, and probably add .DS_Store to .gitignore

**Comment by imjasonh** [comment]
Can you fix the build errors so we can see if any other tests are affected?

```
Error: pkg/v1/local/puller.go:63:21: undefined: errors.ErrUnsupported
```

File: pkg/v1/local/layer.go (hunk 1 above)
Comment (thesayyn) [reply pkg/v1/local/layer.go:1]: Not really, could go into layout package.

**Comment by thesayyn** [comment]
> Can you fix the build errors so we can see if any other tests are affected?
> 
> ```
> Error: pkg/v1/local/puller.go:63:21: undefined: errors.ErrUnsupported
> ```

Done

File: cmd/crane/cmd/export.go [hunk 2]
```diff
@@ -70,20 +70,14 @@ func NewCmdExport(options *[]crane.Option) *cobra.Command {
...
-					img, err = desc.Schema1()
-					if err != nil {
-						return fmt.Errorf("pulling schema 1 image %s: %w", src, err)
-					}
-				} else {
-					img, err = desc.Image()
-					if err != nil {
-						return fmt.Errorf("pulling Image %s: %w", src, err)
-					}
+				var ok bool
```
Comment (thesayyn) [inline cmd/crane/cmd/export.go:77]: This is wrong. inspect media type to see if it's schema1

File: pkg/v1/layout/pusher.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	}
+	b, desc, err := unpackTaggable(t)
+	if err != nil {
+		return err
+	}
+	if err := lp.path.WriteBlob(desc.Digest, io.NopCloser(bytes.NewBuffer(b))); err != nil {
+		return err
+	}
+
+	if img, ok := mf.(v1.Image); ok {
```
Comment (thesayyn) [inline pkg/v1/layout/pusher.go:131]: TODO: image index

File: pkg/v1/remote/image.go [hunk 4]
```diff
@@ -62,12 +63,40 @@ var _ partial.CompressedImageCore = (*remoteImage)(nil)
...
-
-	return desc.Image()
+	desc, err := Artifact(ref, options...)
+	if err != nil {
+		return nil, err
+	}
+	if img, ok := desc.(v1.Image); ok {
+		return img, nil
+	}
+	if idx, ok := desc.(v1.ImageIndex); ok && opts.platform.String() != "" {
```
Comment (thesayyn) [inline pkg/v1/remote/image.go:77]: Better way to do this?

File: pkg/v1/layout/puller.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	"github.com/google/go-containerregistry/pkg/v1/partial"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	specsv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type puller struct {
+	path Path
+}
+
+func NewPuller(path Path) remote.Puller {
```
Comment (jonjohnsonjr) [inline pkg/v1/layout/puller.go:33]: I think we want to define a new interface in `v1` that both `remote.Puller` and `layout.Source` would implement.

In my mind I'd want:

```
package v1

type Source interface{
 // minimal subset of methods needed to read things
}

type Sink interface{
 // minimal subset of methods needed to write things

}
```

File: pkg/v1/remote/puller.go [hunk 6]
```diff
@@ -24,27 +25,40 @@ import (
 	"github.com/google/go-containerregistry/pkg/v1/types"
 )
 
-type Puller struct {
+type Puller interface {
```
Comment (jonjohnsonjr) [inline pkg/v1/remote/puller.go:28]: We want to keep this a concrete struct.

File: pkg/v1/layout/puller.go (hunk 5 above)
Comment (aw185176) [reply pkg/v1/layout/puller.go:33]: That would remove the dependency between `layout` + `remote` and avoid the need to convert `remote.Puller` into a struct below. Spitballing the set of methods:

```go
type Source interface {
  Layer(...)
  Head(...)
  Get(...)
  Artifact(...)
}

type Sink interface {
  Push(...)
  Upload(...) // Maybe not?
}
```

The verbs are a bit clunky for non-remote implementors, but I don't see a clean way around that.

File: pkg/v1/remote/puller.go (hunk 6 above)
Comment (thesayyn) [reply pkg/v1/remote/puller.go:28]: Agreed, will make the change

File: pkg/v1/layout/puller.go (hunk 5 above)
Comment (thesayyn) [reply pkg/v1/layout/puller.go:33]: Hmm, i can't put source and sink into v1 root folder as it creates cycle between `v1` -> `partial.Artifact` -> `v1`


Would putting this into its own directory be acceptable?

```
package github.com/google/go-containerregistry/cmd/crane
        imports github.com/google/go-containerregistry/cmd/crane/cmd
        imports github.com/google/go-containerregistry/internal/cmd
        imports github.com/google/go-containerregistry/internal/verify
        imports github.com/google/go-containerregistry/pkg/v1
        imports github.com/google/go-containerregistry/pkg/v1/partial
        imports github.com/google/go-containerregistry/pkg/v1: import cycle not allowed
```

Comment (thesayyn) [reply pkg/v1/layout/puller.go:33]: okay i have created the sinksource (name TBD) directory.

File: pkg/v1/remote/puller.go (hunk 6 above)
Comment (thesayyn) [reply pkg/v1/remote/puller.go]: Done.

File: pkg/v1/layout/puller.go (hunk 5 above)
Comment (thesayyn) [reply pkg/v1/layout/puller.go]: This is now ready for another look

Comment (aw185176) [reply pkg/v1/layout/puller.go]: @jonjohnsonjr what do you think about `v1.Artifact` to get around the import cycle?

Comment (thesayyn) [reply pkg/v1/layout/puller.go]: that one also sounds fair, though now v1.Artifact would depend on bunch of partial interfaces.

Comment (thesayyn) [reply pkg/v1/layout/puller.go]: anything i could do to push this over the edge @jonjohnsonjr? I'd really like to land this, as i have some people who's been waiting for it for a long time.

**Pull request closed without merging**